### PR TITLE
NoUnneededFinalMethodFixer - Remove final keyword from private methods

### PR DESCRIPTION
**Pull request by localheinz**
This PR

* [x] adjusts the `NoUnneededFinalMethodFixer` to also remove an unnecessary `final` keyword from `private` methods

### CODE REVIEW

File: tests/Fixer/ClassNotation/NoUnneededFinalMethodFixerTest.php [hunk 1]
```diff
@@ -51,6 +51,16 @@ private function baz() {}
...
 }',
             ],
+            'private-method' => [
+                '<?php
+class Foo {
+    private function baz() {}
+}',
+                '<?php
+class Foo {
+    final private function baz() {}
```
Comment (keradus) [inline tests/Fixer/ClassNotation/NoUnneededFinalMethodFixerTest.php]: `final private` method. kudos for PHP

Comment (localheinz) [reply tests/Fixer/ClassNotation/NoUnneededFinalMethodFixerTest.php]: Hehe, actually had an example of this lying around somewhere in **my** code, that's how I figured I should take a look.

**Comment by keradus** [comment]
tests are failing :(

**Comment by julienfalque** [comment]
Please update the PR to target branch 2.6 :)

**Comment by keradus** [comment]
it's a feature imo

**Comment by keradus** [comment]
@julienfalque 

this PR makes one fixer fix more cases, you call it bugfix (targeting 2.6)

#3089 makes one fixer fix mor cases, you call it feature (targeting master)
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3089#pullrequestreview-64574596

**Comment by julienfalque** [comment]
@keradus I consider this one a bugfix because IMO it's something that should have been part of the initial implementation and is missing, while for #3089 it's not the case. Though I'm fine with considering this one as a new feature as well, your call :)

File: src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php [hunk 2]
```diff
@@ -61,24 +67,33 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens)
...
     {
+        $methodVisibilities = [
+            T_PUBLIC,
+            T_PROTECTED,
+            T_PRIVATE,
+        ];
+
+        if (!$classIsFinal) {
+            $methodVisibilities = [
+                T_PRIVATE,
```
Comment (julienfalque) [inline src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php]: `T_PRIVATE` is always part of the array, I would reverse the condition to initialize it with this value only, and add the others when the class is final.

File: src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php [hunk 3]
```diff
@@ -97,6 +112,12 @@ private function fixClass(Tokens $tokens, $classOpenIndex)
                 continue;
             }
 
+            $nextMeaningfulToken = $tokens[$tokens->getNextMeaningfulToken($index)];
+
+            if (!$nextMeaningfulToken->isGivenKind($methodVisibilities)) {
```
Comment (julienfalque) [inline src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php]: This does not cover the case where the visibility keyword is before `final`.

File: tests/Fixer/ClassNotation/NoUnneededFinalMethodFixerTest.php [hunk 4]
```diff
@@ -51,6 +51,16 @@ private function baz() {}
     final private function baz() {}
 }',
             ],
+            'private-method' => [
```
Comment (julienfalque) [inline tests/Fixer/ClassNotation/NoUnneededFinalMethodFixerTest.php]: Please add the following test case:

```diff
 class Foo {
-    private final function bar() {}
+    private function bar() {}
 }
```

**Comment by localheinz** [comment]
Picked https://github.com/FriendsOfPHP/PHP-CS-Fixer/tree/2.6 as target branch now.

**Comment by julienfalque** [comment]
@localheinz That was actually not needed as @keradus changed the target milestone to `2.7` (current master), which means we consider this a new feature. Sorry for the trouble.

**Comment by localheinz** [comment]
@julienfalque 

No problem!

**Comment by keradus** [comment]
Please rebase and make Travis happy

File: src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php [hunk 5]
```diff
@@ -61,34 +67,31 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens)
...
             $prevToken = $tokens[$tokens->getPrevMeaningfulToken($index)];
-            if (!$prevToken->isGivenKind(T_FINAL)) {
-                $index = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_CURLY_BRACE, $classOpen);
+            $classIsFinal = $prevToken->isGivenKind(T_FINAL);
 
-                continue;
-            }
-
-            $index = $this->fixClass($tokens, $classOpen);
+            $this->fixClass($tokens, $classOpen, $classIsFinal);
```
Comment (keradus) [inline src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php]: why this change? like, without `$index = ...` assignment we don't do the jump, so it's slower to finish the main loop. do we need to not do that jump ?

Comment (localheinz) [reply src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php]: If we jump to the end of the class and ignore the content of methods, we cannot fix anonymous class definitions inside methods. 

Do you have a suggestion, @keradus?

Comment (keradus) [reply src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php]: with that reasoning, keep your change :+1:

**Comment by keradus** [comment]
@julienfalque , you added #3094, which got merged.
could you now verify this PR ?

**Comment by keradus** [comment]
Thank you @localheinz.

**Comment by localheinz** [comment]
Thank you, @julienfalque and @keradus!

**Comment by SpacePossum** [comment]
thanks @localheinz and reviewers :D
